feat: Provide support for Managed Identity authentication

Mimetis opened the pull request:
# Adding Managed Identity to Promitor

## Intro

This Pull Request is all about adding the **Managed Identity** option to **Promitor**, adding a new way to authenticate **Promitor** to **Azure**.
It's not a replacement of the existing options using the **Service Principal**.

The two endpoints using an authentication method should be updated:
- **Scraper** Implementation : **Done**
- **Resource Discovery** Implementation : **Done**

![image](https://user-images.githubusercontent.com/4592555/109826991-a4849880-7c3b-11eb-9381-0709910ec1a0.png)

## NOT Handled by this PR

Helm: The Helm Chat Will Have To Be Updated As Well
Documentation: Create a Walkthrough on How to use Pod Identity & Promitor With Managed Identity

## Constraints

We have two distincts projects to modify to integrated the managed identity support:
- `Promitor.Agents.Scraper`
- `Promitor.Agents.ResourceDiscovery`

The best and most recent SDK to use for authenticating a Service Principal or a Managed Identity is `Azure.Identity`.

The `Promitor.Agents.Scraper` uses the Fluent Azure SDK that is not yet compatible with `Azure.Identity` (see [this issue for more info](https://github.com/Azure/azure-sdk-for-net/issues/9262#issuecomment-580444940))
The `Promitor.Agents.ResourceDiscovery` uses the Azure Management SDK that is compatible with `Azure.Identity`.

Until we reach a point where both SDKs are compatible with `Azure.Identity`, we'll need to implement two differents authentication solution for each project.

## Server Options

To be able to use Managed Identity, we need to differentiate the identity options, between a **Service Principal** using one Guid and one Password and a **Managed Identity** using a single Guid for a **User Managed Identity** and nothing when using a **Service Managed Identity**.

As we are in a *"Configuration as Code"* motion, we have added a new section in the `runtime.yaml` configuration called `authentication`.
This new `authentication` section has a first property `mode`:

``` yaml
server:
  httpPort: 88
authentication:  
  mode: managedIdentity
metricSinks:
  ...
telemetry:
  ...
azureMonitor:
  ...
```

`authentication.mode` options can be:
- `servicePrincipal`: The server will use a **Service Principal** authentication method and will check the environment variable `PROMITOR_AUTH_APPID` and `PROMITOR_AUTH_APPKEY`.
- `managedIdentity` : Ther server will use a **Managed Identity** authentication method and will check the environment variable `PROMITOR_AUTH_UAMI`.

> If nothing is mentioned in the `authentication` section, the default authentication mode will be `servicePrincipal`.

## Pod Configuration

Once the server is configured to use either a **Managed Identity** or a **Service Principal**, the Pod should be configured with the corresponding environment variables:

1) Using a **User Managed Identity**:

``` yaml
containers:
  - name: promitor-agent-scraper
    image: spertus.azurecr.io/promitor-agent-scraper:latest
    ports:
      - name: http
        containerPort: 88
        protocol: TCP
    env:
      - name: PROMITOR_AUTH_UAMI
        value: "xxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxx"
```

2) Using a **System Managed Identity**:

``` yaml
containers:
  - name: promitor-agent-scraper
    image: spertus.azurecr.io/promitor-agent-scraper:latest
    ports:
      - name: http
        containerPort: 88
        protocol: TCP
    env:
```
> When using the `managedIdentity` mode, the environment variable `PROMITOR_AUTH_UAMI` can be omitted. Without any value the **Service Managed Identity** affiliated with the underline VM will be used.

3) Using the default **Service Principal**:

``` yaml
containers:
  - name: promitor-agent-scraper
    image: spertus.azurecr.io/promitor-agent-scraper:latest
    ports:
      - name: http
        containerPort: 88
        protocol: TCP
    env:
      - name: PROMITOR_AUTH_APPID
        value: "xxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxx"
      - name: PROMITOR_AUTH_APPKEY
        value: "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
```

Relates to https://github.com/tomkerkhove/promitor/issues/444

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/tomkerkhove/promitor?pullRequest=1544) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/tomkerkhove/promitor?pullRequest=1544) before we can accept your contribution.<br/><hr/>**Sébastien Pertus** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/tomkerkhove/promitor?pullRequest=1544) it.</sub>

tomkerkhove commented on src/Promitor.Agents.Scraper/.config/dotnet-tools.json:
```diff
@@ -0,0 +1,12 @@
+{
```
Can you remove this file please?

Mimetis replied:
Done ! God Damn Visual Studio sometimes !

tomkerkhove replied:
Thanks and no worries!

tomkerkhove commented on src/Promitor.Docker.dcproj:
```diff
@@ -10,9 +10,6 @@
     <DockerServiceName>promitor.scraper</DockerServiceName>
   </PropertyGroup>
   <ItemGroup>
-    <None Include="docker-compose.vs.debug.yml">
```
Please don't remove this

tomkerkhove replied:
Or is it because it breaks your local experience? if so we should check to make it optional

Mimetis replied:
Done !

tomkerkhove replied:
Thanks!

tomkerkhove commented on src/docker-compose.yml:
```diff
@@ -12,7 +13,8 @@ services:
       - ./../config/promitor/resource-discovery/runtime.yaml:/config/runtime.yaml
   promitor.agents.scraper:
-    image: ghcr.io/tomkerkhove/promitor-agent-scraper
+    #image: ghcr.io/tomkerkhove/promitor-agent-scraper
```
Let's rever this please

Mimetis replied:
Done !

tomkerkhove replied:
Thanks!

tomkerkhove commented on src/docker-compose.yml:
```diff
@@ -2,7 +2,8 @@ version: '3.4'
   promitor.agents.resourcediscovery:
-    image: ghcr.io/tomkerkhove/promitor-agent-resource-discovery
+    #image: ghcr.io/tomkerkhove/promitor-agent-resource-discovery
+    image: spertus.azurecr.io/promitor-agent-resource-discovery
```
Let's rever this please

Mimetis replied:
Done !

tomkerkhove replied:
Thanks!

tomkerkhove reviewed (commented):
Thanks for the PR and thorough analysis!

I'm still more of a fan for adding `authentication.mode` to the runtime config so that it's explicit in the authentication mechanics that are requested and would allow for cleaner code as well.

Then people could use:
- `ServicePrinciple`
- `ManagedIdentity`

You could even go further and have `authentication.managedIdentity.type` which can be `SystemAssigned` or `UserAssigned`. By doing this, it's clear for people what they want, but also for others when just reading the configs in their git repo.

Would you mind changing the approach to that please?

Also, I'm not a fan of abbreviations because not everybody knows what `MI`, would you mind renaming it to `ManagedIdentity` please?

Mimetis commented:
Hey @tomkerkhove , @jcorioland 

I have updated the PR with the implementation for **Resource Discovery**.
The authentication mode is now part of the `server` configuration.
I have updated the PR first message with the new configuration, and I tried to make it as concise as possible.

If you guys wants a demo, let me know, we may be able to make a short teams meeting to show case it !

tomkerkhove commented:
> Thanks for the PR and thorough analysis!
> 
> I'm still more of a fan for adding `authentication.mode` to the runtime config so that it's explicit in the authentication mechanics that are requested and would allow for cleaner code as well.
> 
> Then people could use:
> 
> * `ServicePrinciple`
> * `ManagedIdentity`
> 
> You could even go further and have `authentication.managedIdentity.type` which can be `SystemAssigned` or `UserAssigned`. By doing this, it's clear for people what they want, but also for others when just reading the configs in their git repo.
> 
> Would you mind changing the approach to that please?
> 
> Also, I'm not a fan of abbreviations because not everybody knows what `MI`, would you mind renaming it to `ManagedIdentity` please?

Just a small change request, can you align with the above suggestion please ☝

For example, it shouldn't be under `server` and it should be a mode, not the authentication object itself in case we need to add more in the future.

Sorry about that.

Mimetis commented:
You prefer having something like

``` yaml
server:
  httpPort: 88
authentication:
  mode: managedIdentity
```
instead of 
``` yaml
server:
  httpPort: 88
  authentication: managedIdentity
```
This is something I've considered, but in a sense, it seemed to me that the authentication is really tight to the Server itself, since it's the server who needs to authenticate itself to the Azure ressources.
That's why I chose to add the authentication as a new option to the `server` node.

But If you prefer to have a new node, that's perfectly fine for me.

tomkerkhove commented:
Let me know if you are ready and I'd be happy to review it!

Mimetis commented:
I think it's ok for a first review @tomkerkhove :)

tomkerkhove commented on config/promitor/resource-discovery/runtime.yaml:
```diff
@@ -1,5 +1,7 @@
 server:
   httpPort: 88
+authentication:
+  mode: managedIdentity
```
```suggestion
  mode: ManagedIdentity
```

Mimetis replied:
I used to use camel casing in Yaml configuration, but we can do it the way you want

tomkerkhove replied:
It's ok, as long as both are supported :D

tomkerkhove commented on config/promitor/scraper/runtime.yaml:
```diff
@@ -1,5 +1,7 @@
 server:
   httpPort: 88
+authentication:
+  mode: managedIdentity
```
```suggestion
  mode: ManagedIdentity
```

tomkerkhove commented on src/Promitor.Agents.ResourceDiscovery/Graph/Authentication.cs:
```diff
@@ -1,20 +1,55 @@
     public static class Authentication
     {
-        public static async Task<ServiceClientCredentials> GetServiceClientCredentialsAsync(string resource, string clientId, string clientSecret, string tenantId)
+        //public static async Task<ServiceClientCredentials> GetServiceClientCredentialsAsync(string resource, string clientId, string clientSecret, string tenantId)
```
If you no longer need this, please remove it

Mimetis replied:
Oh, I just wanted to be sure nothing were lost (just in case) but obviously, this is not used anymore

tomkerkhove replied:
Kill it with 🔥 Thanks though :D

tomkerkhove commented on src/Promitor.Core/EnvironmentVariables.cs:
```diff
@@ -11,6 +11,7 @@ public class Authentication
         {
             public const string ApplicationId = "AUTH_APPID";
             public const string ApplicationKey = "AUTH_APPKEY";
+            public const string ManagedIdentityId = "AUTH_UAMI";
```
Let's use this to be more clear?
```suggestion
            public const string IdentityId = "AUTH_MANAGED_IDENTITY_ID";
```

Mimetis replied:
Well, UAMI is the acronym used everywhere (you will often find it on stackoverflow for instance) when we are referencing a User Assigned Managed Identity.
If you want to have a full name identifier, I would replace it with "AUTH_USER_ASSIGNED_MANAGED_IDENTITY_ID"
Choose the one you prefer :)

tomkerkhove replied:
I know but it's a bit too long so I went for this, don't want to overcomplicate it :D

tomkerkhove commented on src/Promitor.Core.Scraping/Promitor.Core.Scraping.csproj:
```diff
@@ -26,6 +26,7 @@
   </ItemGroup>
 
   <ItemGroup>
+    <ProjectReference Include="..\Promitor.Agents.Core\Promitor.Agents.Core.csproj" />
```
What do you need from agent core here? Typically it would be the other way around.

Mimetis replied:
The configuration section, where we have `AuthenticationConfiguration` and `AuthenticationMode`

tomkerkhove replied:
I'll merge it as is and quickly check it out

tomkerkhove commented on src/Promitor.Integrations.AzureMonitor/Promitor.Integrations.AzureMonitor.csproj:
```diff
@@ -22,6 +22,7 @@
   </ItemGroup>
 
   <ItemGroup>
+    <ProjectReference Include="..\Promitor.Agents.Core\Promitor.Agents.Core.csproj" />
```
What is it that you need from agent core here? Normally it should need the reference

tomkerkhove commented on src/Promitor.Integrations.AzureMonitor/AzureMonitorClient.cs:
```diff
@@ -217,12 +225,25 @@ private MetricValue GetMostRecentMetricValue(string metricName, TimeSeriesElemen
+        private IAzure CreateAzureClient(AzureEnvironment azureCloud, string tenantId, string subscriptionId, AuthenticationMode authenticationMode, string managedIdentityId, string applicationId, string applicationSecret, IOptions<AzureMonitorLoggingConfiguration> azureMonitorLoggingConfiguration, ILoggerFactory loggerFactory, MetricSinkWriter metricSinkWriter, IRuntimeMetricsCollector metricsCo ...
         {
-            var credentials = _azureCredentialsFactory.FromServicePrincipal(applicationId, applicationSecret, tenantId, azureCloud);
+            AzureCredentials credentials;
+
+            if (authenticationMode == AuthenticationMode.ManagedIdentity && !string.IsNullOrEmpty(managedIdentityId))
+            {
+                credentials = _azureCredentialsFactory.FromUserAssigedManagedServiceIdentity(managedIdentityId, MSIResourceType.VirtualMachine, azureCloud, tenantId);
+            }
+            else if (authenticationMode == AuthenticationMode.ManagedIdentity)
+            {
+                credentials = _azureCredentialsFactory.FromSystemAssignedManagedServiceIdentity(MSIResourceType.VirtualMachine, azureCloud, tenantId);
+            }
+            else
+            {
+                credentials = _azureCredentialsFactory.FromServicePrincipal(applicationId, applicationSecret, tenantId, azureCloud);
+            }
```
I would move this to a dedicated method
```suggestion
            AzureCredentials credentials = CreateAzureAuthentication(azureCloud, authenticationMode, tenantId, managedIdentityId);
```

Mimetis replied:
As you want !

tomkerkhove commented on src/Promitor.Integrations.AzureMonitor/AzureMonitorClient.cs:
```diff
@@ -33,22 +34,29 @@ public class AzureMonitorClient
         {
             Guard.NotNullOrWhitespace(tenantId, nameof(tenantId));
+
+            if (authenticationMode == AuthenticationMode.ServicePrincipal)
```
Maybe we should move the guarding to CreateAzureAuthentication?

Mimetis replied:
If we create the method `CreateAzureAuthentication`, I guess that could be a good solution

tomkerkhove replied:
Let's do that!

tomkerkhove commented on src/Promitor.Integrations.AzureMonitor/AzureMonitorClient.cs:
```diff
@@ -217,12 +225,25 @@ private MetricValue GetMostRecentMetricValue(string metricName, TimeSeriesElemen
-            var credentials = _azureCredentialsFactory.FromServicePrincipal(applicationId, applicationSecret, tenantId, azureCloud);
+            AzureCredentials credentials;
+
+            if (authenticationMode == AuthenticationMode.ManagedIdentity && !string.IsNullOrEmpty(managedIdentityId))
```
Instead of making this assumption, what if `AuthenticationMode` uses `UserAssigedManagedIdentity` and `SystemAssigedManagedIdentity`?

We can then simplify the code (switch/case) and improved checks?

Mimetis replied:
The difference between UAMI and SAMI can be quite complex for a user that is not used to use them.
That's why I prefer to make the assumption regarding if user has provided the GUID.
And the "trick" is that aad podidentity "transforms" a UAMI into a virtual SAMI, and that could be disturbing at some point.
Once again, if you prefer add the 2 options, it's ok for me, but I think it will lead to more errors if the configuration is not set correctly by the user

tomkerkhove replied:
I'm thorn on this. I'll think about it. On one hand it's explicit with multiple options, on the other hand it requires more knowledge indeed.

tomkerkhove replied:
Let's split them, I'll do docs on it which should make it clear 🤷‍♂️

tomkerkhove commented on src/Promitor.Agents.Scraper/AzureMonitorClientFactory.cs:
```diff
@@ -43,17 +44,29 @@ public AzureMonitorClient CreateIfNotExists(AzureEnvironment cloud, string tenan
+            // If not, we should use a default value
+            if (authenticationConfiguration == null)
+            {
+                authenticationConfiguration = new AuthenticationConfiguration();
```
We might want to add a logging statement here that they should explicitly configure it through a warning?

Mimetis replied:
Oh, indeed, a log warning can be a good idea. Or maybe just a log information since it will still working even if the user does not specify the authentication mode he wants to use.
Eventually this warning (or info) will concern only the user that will upgrade Promitor to the last version, with an already in place configuration

tomkerkhove replied:
Yes, but I'd use a warning since it's higher than info so people will certainly see it - Make sense?

tomkerkhove commented on src/Promitor.Agents.ResourceDiscovery/Graph/AzureResourceGraph.cs:
```diff
@@ -105,8 +140,20 @@ private async Task<TResponse> InteractWithAzureResourceGraphAsync<TResponse>(str
+                                {
+                                    unauthorizedException = new UnauthorizedException(QueryManagedIdentityId, targetSubscriptions);
+                                    _logger.LogCritical(unauthorizedException, "Unable to query Azure Resource Graph using the Managed Identity");
+                                    _logger.LogCritical("Azure Resource Graph is using a Managed Identity to authenticate");
```
We already have a log entry for this
```suggestion
```

Mimetis replied:
Indeed !

tomkerkhove commented on src/Promitor.Agents.ResourceDiscovery/Graph/AzureResourceGraph.cs:
```diff
@@ -105,8 +140,20 @@ private async Task<TResponse> InteractWithAzureResourceGraphAsync<TResponse>(str
+                                {
+                                    unauthorizedException = new UnauthorizedException(QueryApplicationId, targetSubscriptions);
+                                    _logger.LogCritical(unauthorizedException, "Unable to query Azure Resource Graph using the Service Principal");
+                                    _logger.LogCritical("Azure Resource Graph is using a Managed Identity to authenticate");
```
We already have a log entry for this
```suggestion
```

tomkerkhove commented on src/Promitor.Agents.ResourceDiscovery/Graph/AzureResourceGraph.cs:
```diff
@@ -105,8 +140,20 @@ private async Task<TResponse> InteractWithAzureResourceGraphAsync<TResponse>(str
                             {
-                                var unauthorizedException = new UnauthorizedException(QueryApplicationId, targetSubscriptions);
-                                _logger.LogCritical(unauthorizedException, "Unable to query Azure Resource Graph");
+                                UnauthorizedException unauthorizedException;
```
Maybe we should add the type of auth to the exception?

Mimetis replied:
Indeed ! :)

tomkerkhove commented on src/Promitor.Integrations.AzureMonitor/RequestHandlers/AzureResourceManagerThrottlingRequestHandler.cs:
```diff
@@ -46,11 +54,19 @@ public AzureResourceManagerThrottlingRequestHandler(string tenantId, string subs
             _metricSinkWriter = metricSinkWriter;
             _metricsCollector = metricsCollector;
 
+            string id = applicationId;
+
+            if (authenticationMode == AuthenticationMode.ManagedIdentity)
+            {
+                id = string.IsNullOrEmpty(managedIdentityId) ? "system-assigned-identity" : managedIdentityId;
+            }
```
Let's move this to a method
```suggestion
            DetermineApplicationId(...)
```

tomkerkhove commented on src/Promitor.Integrations.AzureMonitor/RequestHandlers/AzureResourceManagerThrottlingRequestHandler.cs:
```diff
@@ -46,11 +54,19 @@ public AzureResourceManagerThrottlingRequestHandler(string tenantId, string subs
                 {"subscription_id", subscriptionId},
-                {"app_id", applicationId}
+                {"authentication", authenticationMode.ToString()},
+                {"authentication_id", id},
```
I think we should keep using `app_id`
```suggestion
                {"app_id", id},
```

tomkerkhove commented on src/Promitor.Integrations.AzureMonitor/RequestHandlers/AzureResourceManagerThrottlingRequestHandler.cs:
```diff
@@ -46,11 +54,19 @@ public AzureResourceManagerThrottlingRequestHandler(string tenantId, string subs
                 {"tenant_id", tenantId},
                 {"subscription_id", subscriptionId},
-                {"app_id", applicationId}
+                {"authentication", authenticationMode.ToString()},
```
Not sure if this is valuable info since this is applicable for all metrics and implementation detail?

Mimetis replied:
Do you want to log only tenant_id and subscription_id (and remove auth) ?

tomkerkhove replied:
Yeah, maybe that makes more sense but that's a breaking change so would like to keep `app_id` for now

tomkerkhove commented on src/Promitor.Agents.Core/Validation/Steps/AzureAuthenticationValidationStep.cs:
```diff
@@ -21,16 +22,41 @@ public AzureAuthenticationValidationStep(IConfiguration configuration, ILogger<A
+            // If not, we should use a default value
+            if (authenticationConfiguration == null)
+            {
+                authenticationConfiguration = new AuthenticationConfiguration();
```
Let's log a warning here that folks should specify it maybe?

tomkerkhove commented on src/Promitor.Agents.Core/Validation/Steps/AzureAuthenticationValidationStep.cs:
```diff
@@ -21,16 +22,41 @@ public AzureAuthenticationValidationStep(IConfiguration configuration, ILogger<A
         {
-            var applicationId = _configuration.GetValue<string>(EnvironmentVariables.Authentication.ApplicationId);
-            if (string.IsNullOrWhiteSpace(applicationId))
+            var authenticationConfiguration = _configuration.GetSection("authentication").Get<AuthenticationConfiguration>();
```
Some nits for this method:
- Can you add `{}` to the ifs, even if it's a oneliner?
- You don't need to prefix it with `this.`

Mimetis replied:
Ah ah, I'm used to use `this` everytime :) 
I did too much TypeScript I guess !

tomkerkhove replied:
Haha that's ok, I'm just being a pain :D

tomkerkhove commented on src/Promitor.Agents.ResourceDiscovery/Graph/Authentication.cs:
```diff
@@ -1,20 +1,55 @@
+                tokenCredential = new DefaultAzureCredential();
+            }
+
+            if (!resource.ToLowerInvariant().EndsWith("/.default"))
```
Can you elaborate why we need this?

Mimetis replied:
When you reaching an endpoint, using a impersonate identity, the only endpoint available is always '/.default' 
MSAL add the './default' string to your resource request behind the scene.
We have to do it here, since we are at a lower level (and we are not using MSAL; by the way)

tomkerkhove replied:
That's ok, thanks for explaining! Would you mind adding a comment with that please?

tomkerkhove reviewed (commented):
Added a few comments but we're almost there.

After that I'll create a feature branch to which we can merge and I'll work on the rest such as docs and minor nits.

Thanks a ton!

tomkerkhove commented:
If you sign the CLA @Mimetis and clarify [this comment](https://github.com/tomkerkhove/promitor/pull/1544#discussion_r590492352) then I'm happy to pick up the rest if you prefer?

Mimetis commented:
Hey @tomkerkhove 
I can make the comments and let you implement everything the way you prefer.
I think it will be way faster.
BUT if you don't have the bandwidth, I'm available these days to do it, let me know.

Mimetis commented:
@tomkerkhove I've commented everything.
Let me know if it's ok for you and if you want me to implement everything.
If you prefer to do it by yourself, I guess you've all the info you need to take all the design / naming / methods decisions

Mimetis commented:
by the way, i've already signed the CLA, some days ago

![image](https://user-images.githubusercontent.com/4592555/110761271-d1463a80-824f-11eb-859c-fe92d3336fe9.png)

tomkerkhove commented:
> @tomkerkhove I've commented everything.
> Let me know if it's ok for you and if you want me to implement everything.
> If you prefer to do it by yourself, I guess you've all the info you need to take all the design / naming / methods decisions

No it's ok if you want to do it, thanks a ton!!!



> by the way, i've already signed the CLA, some days ago
> 
> ![image](https://user-images.githubusercontent.com/4592555/110761271-d1463a80-824f-11eb-859c-fe92d3336fe9.png)

Huh, I'll check the tooling ,thanks for commenting!

tomkerkhove commented:
I've checked the CLA thingy and turns out this is the reason:
![image](https://user-images.githubusercontent.com/4345663/111118028-8f2b3a80-8568-11eb-88d4-7f4df8d465b1.png)

Can you check your local Git settings please? We can merge it as is since you've signed it though.

tomkerkhove reviewed (approved):
Thank you for your contribution, @Mimetis - Super helpful!

tomkerkhove commented:
I'm going to do some minor refactoring - Can you live with the fact that `AUTH_APP_ID` & `AUTH_MANAGED_IDENTITY_ID` would be merged into `authentication.identityId`?

Mimetis commented:
As you wish ! It's your library, dude ;)

Mimetis commented:
Just need the final result to work on the walkthrough and the Promitor charts as well :)

tomkerkhove commented:
> As you wish ! It's your library, dude ;)

But I'm building it for the users, not myself so wondering if that makes sense :)

Let's go that route, I'll make the changes.

Mimetis commented:
Does it mean that we will have the `app_id` in the `runtime.yaml` and the `app_key` in the environment variables ?

tomkerkhove commented:
Yes to align with new approach since it's not a secret

Mimetis commented:
And what about the compatibility with previous version ?

tomkerkhove commented:
If it's not in the config, then we assume the env var is there (but new Helm chart should use the YAML). Make sense?

Mimetis commented:
Well, makes sense if we agree that we will need to check both places.

Checking the `authentication` node and if there is nothing, checking the environment variables:

``` cs

// trying to get app_id:
var authenticationConfiguration = _configuration.GetSection("authentication").Get<AuthenticationConfiguration>();
var applicationId = authenticationConfiguration.IdentityId

// trying to get app_id from env variables
var applicationId = _configuration.GetValue<string>(EnvironmentVariables.Authentication.ApplicationId); 
// PROMITOR_AUTH_APPID
```
If it is something you are comfortable with, that's perfect :)

tomkerkhove commented:
Most definately yes, that's the goal indeed - Thanks for confirming it's ok!
